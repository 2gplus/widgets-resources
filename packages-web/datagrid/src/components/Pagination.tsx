import { createElement, ReactElement, Dispatch, SetStateAction } from "react";

export interface PaginationProps {
    page: number;
    setPage: Dispatch<SetStateAction<number>>;
    hasMoreItems: boolean;
}

export interface ClientSidePaginationProps {
    gotoPage: (page: number) => void;
    previousPage: () => void;
    nextPage: () => void;
    page: number;
    pageOptions: number[];
    canPreviousPage: boolean;
    canNextPage: boolean;
}

export function Pagination({ page, setPage, hasMoreItems }: PaginationProps): ReactElement {
    return (
        <div className="pagination">
            <button className="btn" onClick={() => setPage(0)} disabled={page === 0}>
                <span className="glyphicon glyphicon-step-backward" />
            </button>
            <button className="btn" onClick={() => setPage(prev => prev - 1)} disabled={page === 0}>
                <span className="glyphicon glyphicon-backward" />
            </button>
            <div className="paging-status">Page {page + 1}</div>
            <button className="btn" onClick={() => setPage(prev => prev + 1)} disabled={!hasMoreItems}>
                <span className="glyphicon glyphicon-forward" />
            </button>
        </div>
    );
}

export function ClientSidePagination({
    gotoPage,
    previousPage,
    page,
    nextPage,
    pageOptions,
    canNextPage,
    canPreviousPage
}: ClientSidePaginationProps): ReactElement {
    const lastPage = pageOptions?.length - 1;
    return (
        <div className="pagination">
            <button className="btn" onClick={() => gotoPage(0)} disabled={page === 0}>
                <span className="glyphicon glyphicon-step-backward" />
            </button>
            <button className="btn" onClick={() => previousPage()} disabled={!canPreviousPage}>
                <span className="glyphicon glyphicon-backward" />
            </button>
            <div className="paging-status">
                Page {page + 1} of {pageOptions.length}
            </div>
            <button className="btn" onClick={() => nextPage()} disabled={!canNextPage}>
                <span className="glyphicon glyphicon-forward" />
            </button>
            <button className="btn" onClick={() => gotoPage(lastPage)} disabled={page === lastPage}>
                <span className="glyphicon glyphicon-step-forward" />
            </button>
        </div>
    );
}
