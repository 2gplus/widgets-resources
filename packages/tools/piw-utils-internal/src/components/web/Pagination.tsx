import { createElement, Dispatch, HTMLAttributes, ReactElement, SetStateAction, useEffect, useState } from "react";

export interface PaginationProps {
    canNextPage: boolean;
    canPreviousPage: boolean;
    gotoPage: (page: number) => void;
    nextPage: () => void;
    numberOfItems?: number;
    page: number;
    pageSize: number;
    previousPage: () => void;
    setPaginationIndex?: Dispatch<SetStateAction<number>>;
    labelNextPage?: string;
    labelPreviousPage?: string;
    labelFirstPage?: string;
    labelLastPage?: string;
    labelPagination?: string;
    labelPagingStatus?: string;
}

export function Pagination(props: PaginationProps): ReactElement | null {
    const numberOfPages =
        props.numberOfItems !== undefined ? Math.ceil(props.numberOfItems / props.pageSize) : undefined;
    const lastPage = numberOfPages !== undefined ? numberOfPages - 1 : 0;
    const hasLastPage = numberOfPages !== undefined;
    const initialItem = props.page * props.pageSize + 1;
    const lastItem =
        props.canNextPage || !props.numberOfItems ? (props.page + 1) * props.pageSize : props.numberOfItems;
    const setPageIndex = (page: number): void => {
        if (props.setPaginationIndex) {
            props.setPaginationIndex(page);
        }
    };
    let pagination = "";

    if (props.numberOfItems === 0) {
        return null;
    }

    const [paginationStatus, setPaginationStatus] = useState<string>(
        `${initialItem} "to" ${lastItem} ${
            hasLastPage ? ` "of" ${props.numberOfItems ?? (numberOfPages ?? 1) * props.pageSize}` : ""
        }`
    );
    useEffect(() => {
        const metaModel = async (): Promise<void> => {
            const response = await fetch(`metamodel.json?${mx.server.getCacheBust()}`);
            const metaModel = await response.json();
            const languageId = metaModel.languages.indexOf(mx.session.sessionData.locale.code);
            pagination = metaModel.systemTexts["mendix.lib.MxDataSource.status"][languageId];
            setPaginationStatus(
                pagination
                    .replace("{1}", String(initialItem))
                    .replace("{2}", String(lastItem))
                    .replace("{3}", String(props.numberOfItems ?? (numberOfPages ?? 1) * props.pageSize))
            );
        };
        metaModel().catch(console.error);
    }, [props.numberOfItems, props.pageSize, numberOfPages, props.page]);

    return (
        <div aria-label={props.labelPagination ?? "Pagination"} className="pagination-bar" role="pagination">
            <button
                className="btn pagination-button"
                disabled={props.page === 0}
                {...getEvents(() => {
                    props.gotoPage(0);
                    setPageIndex(0);
                })}
                aria-label={props.labelFirstPage ?? "Go to first page"}
            >
                <span aria-hidden className="glyphicon glyphicon-step-backward" />
            </button>
            <button
                className="btn pagination-button"
                disabled={!props.canPreviousPage}
                {...getEvents(() => {
                    props.previousPage();
                    setPageIndex(props.page - 1);
                })}
                aria-label={props.labelPreviousPage ?? "Go to previous page"}
            >
                <span aria-hidden className="glyphicon glyphicon-backward" />
            </button>
            <span className="sr-only sr-only-focusable">
                {props.labelPagingStatus ?? "Currently showing"} {paginationStatus}
            </span>
            <div aria-hidden className="paging-status">
                {paginationStatus}
            </div>
            <button
                aria-label={props.labelNextPage ?? "Go to next page"}
                className="btn pagination-button"
                disabled={!props.canNextPage}
                {...getEvents(() => {
                    props.nextPage();
                    setPageIndex(props.page + 1);
                })}
            >
                <span aria-hidden className="glyphicon glyphicon-forward" />
            </button>
            {hasLastPage && (
                <button
                    aria-label={props.labelLastPage ?? "Go to last page"}
                    className="btn pagination-button"
                    disabled={props.page === lastPage}
                    {...getEvents(() => {
                        props.gotoPage(lastPage);
                        setPageIndex(lastPage);
                    })}
                >
                    <span aria-hidden className="glyphicon glyphicon-step-forward" />
                </button>
            )}
        </div>
    );
}

function getEvents(action: () => void): Partial<HTMLAttributes<HTMLButtonElement>> {
    return {
        onClick: action,
        onKeyDown: e => {
            if (e.key === " " || e.key === "Enter") {
                e.preventDefault();
                e.stopPropagation();
                action();
            }
        }
    };
}
