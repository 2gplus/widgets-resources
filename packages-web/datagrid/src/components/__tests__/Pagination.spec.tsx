import { shallow } from "enzyme";
import { createElement } from "react";
import { ClientSidePagination, ClientSidePaginationProps, Pagination, PaginationProps } from "../Pagination";

describe("Pagination", () => {
    it("renders the structure correctly", () => {
        const component = shallow(<Pagination {...mockPaginationProps()} />);

        expect(component).toMatchSnapshot();
    });
});

describe("Client Side Pagination", () => {
    it("renders the structure correctly", () => {
        const component = shallow(<ClientSidePagination {...mockClientSideProps()} />);

        expect(component).toMatchSnapshot();
    });
});

function mockPaginationProps(): PaginationProps {
    return {
        hasMoreItems: false,
        setPage: jest.fn(),
        page: 0
    };
}

function mockClientSideProps(): ClientSidePaginationProps {
    return {
        canPreviousPage: true,
        canNextPage: true,
        pageOptions: [1, 2, 3, 4, 5],
        nextPage: jest.fn(),
        previousPage: jest.fn(),
        gotoPage: jest.fn(),
        page: 0
    };
}
