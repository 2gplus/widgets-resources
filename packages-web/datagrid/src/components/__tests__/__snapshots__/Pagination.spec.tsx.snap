// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders the structure correctly 1`] = `
<div
  className="pagination"
>
  <button
    className="btn"
    disabled={true}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-step-backward"
    />
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-backward"
    />
  </button>
  <div
    className="paging-status"
  >
    Page 
    1
     
    of 5
  </div>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-forward"
    />
  </button>
  <button
    className="btn"
    disabled={false}
    onClick={[Function]}
  >
    <span
      className="glyphicon glyphicon-step-forward"
    />
  </button>
</div>
`;
